### PR TITLE
Added Check for 'forbids name from content'

### LINKED ISSUE #50: Forbidden 'Name From Content'
See #32 

- `accname` fails many WPTs because it does not ignore elements with `role="menu"`.
- As it turns out, there is a list of node characteristics that should prevent or forbid that node from taking name from content. See: https://lists.w3.org/Archives/Public/public-aria/2017Jun/0057.html
- Adding a check for nodes that are forbidden from taking name from content to step 2F would result in `accname` passing at least 10 more WPTs.

### PR DESCRIPTION
**Pull request by OisinNolan**
- Added a check for nodes that _forbid name from content_ to step 2F.
- Added a test ([an accname WPT](http://wpt.live/accname/name_checkbox-label-embedded-menu-manual.html)) for this.

Closes #50 

- Some exceptions to this rule to be added soon. See the list 'lst3' in this thread:
https://lists.w3.org/Archives/Public/public-aria/2017Jun/0057.html for more detail.

### CODE REVIEW

File: src/lib/rule2F.ts [hunk 1]
```diff
@@ -138,6 +138,93 @@ function allowsNameFromContent(elem: HTMLElement): boolean {
 
 export const TEST_ONLY = {allowsNameFromContent};
 
+// See https://lists.w3.org/Archives/Public/public-aria/2017Jun/0057.html
+// for discussion of roles & tags that forbid name from content.
```
Comment (AlexLloyd0) [inline src/lib/rule2F.ts:142]: Please explicitly note that this isn't included in 1.1 but as per that thread implementations should check this anyway

File: src/lib/rule2F.ts [hunk 2]
```diff
@@ -138,6 +138,93 @@ function allowsNameFromContent(elem: HTMLElement): boolean {
...
+const NO_NAME_FROM_CONTENT_ROLES = [
```
Comment (AlexLloyd0) [inline src/lib/rule2F.ts:143]: I think `NAME_FROM_CONTENT_FORBIDDEN_ROLES` or `NEVER_NAME_FROM_CONTENT_ROLES` makes it more clear (especially when you add conditional name from content roles)

File: src/lib/rule2F.ts [hunk 3]
```diff
@@ -138,6 +138,93 @@ function allowsNameFromContent(elem: HTMLElement): boolean {
...
+  'tbody',
+  'tfoot',
+  'fieldset',
+];
+
+/**
+ * Checks if 'elem' is forbidden from allowing 'name from content'
+ * @param elem - element whose text alternative is being computed
+ */
+function forbidsNameFromContent(elem: HTMLElement): boolean {
```
Comment (AlexLloyd0) [inline src/lib/rule2F.ts:214]: From the names I'd expect `allowsNameFromContent(el) === !forbidsNameFromContent(el);`. How about combining this logic into `allowsNameFromContent`?

File: src/lib/rule2F.ts [hunk 4]
```diff
@@ -138,6 +138,93 @@ function allowsNameFromContent(elem: HTMLElement): boolean {
...
+  const explicitRole = elem.getAttribute('role')?.trim().toLowerCase() ?? '';
```
Comment (AlexLloyd0) [inline src/lib/rule2F.ts:215]: Should this also include implicit roles?

**Review by AlexLloyd0 — CHANGES_REQUESTED** [review]
Please add a note to this PR description that the more complicated conditional case will come later :)

File: src/lib/rule2F.ts (hunk 3 above)
Comment (OisinNolan) [reply src/lib/rule2F.ts:214]: I think the naming might be misleading here -- I could possibly change the wording to 'roleAllowsNameFromContent' / 'roleForbidsNameFromContent' to clarify.

The idea here being that in 2F we might use the content (subtree) of the current node as part of the name computation in the following cases:
- its role allows name from content
- it is part of a label
- it is descendent of a node whose accessible name is being computed

But if the role forbids name from content, then we shouldn't look at the subtree even if the current node is part of a label etc.

I could combine these into a single function `allowsNameFromContent` that looks at current node role, whether it's part of a label etc. all in the one function? This could even be called `rule2FCondition`.

File: src/lib/rule2F.ts (hunk 4 above)
Comment (OisinNolan) [reply src/lib/rule2F.ts:215]: The implicit roles are handled in the next `if` statement:

```
const elemNodeName = elem.nodeName.toLowerCase();
  if (NEVER_NAME_FROM_CONTENT_TAGS.includes(elemNodeName)) {
    return true;
  }
```

I'll add a comment here to make this clear and more readable.

File: bundle.js [hunk 5]
```diff
@@ -568,14 +568,14 @@ var accname = (function (exports) {
      * @param elem - the function checks if 'elem' allows name from content
      * @return - true if elem allows name from content, false otherwise
      */
-    function allowsNameFromContent(elem) {
+    function roleAllowsNameFromContent(elem) {
```
Comment (AlexLloyd0) [inline bundle.js:571]: The roles from list3 also (sometimes) allows name from content. Maybe add `always` in there?

File: bundle.js [hunk 6]
```diff
@@ -584,15 +584,108 @@ var accname = (function (exports) {
         }
         return false;
     }
+    // See https://lists.w3.org/Archives/Public/public-aria/2017Jun/0057.html
+    // for discussion of roles & tags that forbid name from content.
+    //
+    // *This case is not explicitly included in version 1.1 of the spec, however,
+    // as per the thread linked above we have included it (as have other implementations).
+    const NEVER_NAME_FROM_CONTENT_ROLES = [
```
Comment (AlexLloyd0) [inline bundle.js:592]: nit: Alphabetise these lists so they're easier for readers of the code to parse

File: src/lib/rule2F.ts (hunk 4 above)
Comment (AlexLloyd0) [reply src/lib/rule2F.ts:219]: But the tag name doesn't entirely represent the semantics of  role. For example `<header>` only has a role of `banner` if it's `not a descendant of an article, aside, main, nav or section element, or an element with role=article, complementary, main, navigation or region`... https://www.w3.org/TR/html-aria/#docconformance

For now please just raise an issue, I can add role-finding once my code for that is open source

File: src/lib/compute_text_alternative_test.ts [hunk 7]
```diff
@@ -157,4 +157,25 @@ describe('The computeTextAlternative function', () => {
     const elem = document.getElementById('test')!;
     expect(computeTextAlternative(elem).name).toBe('Flash the screen 2 times.');
   });
+
+  // http://wpt.live/accname/name_checkbox-label-embedded-menu-manual.html
+  it('allows aria-owned nodes to be chained together across multiple nodes', () => {
```
Comment (AlexLloyd0) [inline src/lib/compute_text_alternative_test.ts:162]: update

**Review by AlexLloyd0 — CHANGES_REQUESTED** [review]

File: src/lib/compute_text_alternative_test.ts [hunk 8]
```diff
@@ -157,4 +157,25 @@ describe('The computeTextAlternative function', () => {
     const elem = document.getElementById('test')!;
     expect(computeTextAlternative(elem).name).toBe('Flash the screen 2 times.');
   });
+
+  // http://wpt.live/accname/name_checkbox-label-embedded-menu-manual.html
```
Comment (AlexLloyd0) [inline src/lib/compute_text_alternative_test.ts:161]: How is this link relevant?

File: src/lib/rule2F.ts [hunk 9]
```diff
@@ -109,44 +187,80 @@ function getFunctionCalculatingAllowsNameFromContent(
...
+ * be used in its accesssible name. This is the condition for
+ * rule 2F.
  * @param elem - elem whose text alternative is being computed
  * @param context - additional information about the context of elem
  * @return - whether or not rule 2Fs condition has been satisfied
  */
-function rule2FCondition(elem: HTMLElement, context: Context): boolean {
-  if (allowsNameFromContent(elem)) {
+function allowsNameFromContent(elem: HTMLElement, context: Context): boolean {
+  // focasable elements should allow name from content.
```
Comment (AlexLloyd0) [inline src/lib/rule2F.ts:228]: Should this check be combined with a check for membership of list 3?

> // As an override of list2, conditionally include name from content if current node is focusable, or if the current node matches list3 while the referenced parent node matches list1.

And it looks like the conditions for focusability  are more complicated than `hasAttribute('tabindex')`, see https://stackoverflow.com/questions/1599660/which-html-elements-can-receive-focus and https://html.spec.whatwg.org/multipage/interaction.html#focusable

File: src/lib/compute_text_alternative_test.ts (hunk 8 above)
Comment (OisinNolan) [reply src/lib/compute_text_alternative_test.ts:161]: The unit test on the next lines (162 - 180) is found at this link.

File: src/lib/rule2F.ts (hunk 9 above)
Comment (OisinNolan) [reply src/lib/rule2F.ts:228]: List 3 is handled lines 236-245.

I used the same stackoverflow page you linked as my source for checking whether an element is focusable, I think that 2F handles all of the examples given on this thread:
- `HTMLAnchorElement/HTMLAreaElement with an href` handled by 2F (172-179)
- `HTMLInputElement/HTMLSelectElement/HTMLTextAreaElement/HTMLButtonElement` Inputs, selects and textareas shouldn't take name from content according to the rest of the implementations (checked in the comparison app), as these are handled in special ways (2E). Buttons are handled in 2F line 34.
- `HTMLIFrameElement` Not sure what 'name from content' would look like for an iframe, in any case none of the other implementations allow name from content here.

Could be missing something here but hopefully that covers it :)

File: src/lib/rule2F.ts [hunk 10]
```diff
@@ -226,7 +226,10 @@ function matchesRole(elem: HTMLElement, roleType: RoleType): boolean {
  */
 function allowsNameFromContent(elem: HTMLElement, context: Context): boolean {
   // focasable elements should allow name from content.
-  if (elem.hasAttribute('tabindex')) {
+  if (
+    matchesRole(elem, NEVER_NAME_FROM_CONTENT) &&
+    elem.hasAttribute('tabindex')
+  ) {
     return true;
   }
```
Comment (AlexLloyd0) [inline src/lib/rule2F.ts]: Does moving this to L252 work? Or does L240 catch some elements?

```typescript
  if (matchesRole(elem, NEVER_NAME_FROM_CONTENT)) {
    // focasable elements should allow name from content.
    return elem.hasAttribute('tabindex');
  }
```

**Review by AlexLloyd0 — APPROVED** [review]

File: src/lib/rule2F.ts (hunk 10 above)
Comment (OisinNolan) [reply src/lib/rule2F.ts]: At L252 it wouldn't overwrite list 2 roles, so I think this check has to come first, right?
